Spigot-API-Patches/0317-Adding-enhanced-Material-Item-Color-Type-API.patch

```diff
@@ -0,0 +1,489 @@
++     * @throws IllegalArgumentException if the material has no {@link DyeableMaterial}-group
++     */
++    @NotNull
++    public Material changeColor(@NotNull ColorType newColor) throws IllegalArgumentException {
++            DyeableMaterial material = getDyeableMaterial();
++            if (material != null) {
++                return material.withColor(newColor);
++            } else
++                throw new IllegalArgumentException("The material " + this.name() + " does not have color-support!");
++    }
++    // Paper end
+ }
+diff --git a/src/main/java/org/bukkit/inventory/ItemStack.java b/src/main/java/org/bukkit/inventory/ItemStack.java
+index a7909406e9d54c1ab4789b984ed6b1da50837fce..fdb5a8ba30db16cd8899b0d49f9590be7bf25e43 100644
+--- a/src/main/java/org/bukkit/inventory/ItemStack.java
++++ b/src/main/java/org/bukkit/inventory/ItemStack.java
+@@ -5,6 +5,8 @@ import java.util.LinkedHashMap;
+ import java.util.List; // Paper
+ import java.util.Map;
+ import java.util.Set; // Paper
++import io.papermc.paper.item.ColorType; // Paper
++import io.papermc.paper.item.DyeableMaterial; // Paper
+ import org.apache.commons.lang.Validate;
+ import org.bukkit.Bukkit;
+ import org.bukkit.Material;
+@@ -920,5 +922,43 @@ public class ItemStack implements Cloneable, ConfigurationSerializable, net.kyor
+     public boolean canRepair(@NotNull ItemStack toBeRepaired) {
+         return Bukkit.getUnsafe().isValidRepairItemStack(toBeRepaired, this);
+     }
++
++    /**
++     * Gets the {@link ColorType} of this {@link ItemStack}s-{@link Material}
++     * @return the color of this {@link ItemStack}s-{@link Material}, or null if it has no supported color
++     */
++    @Nullable
++    public ColorType getColorType() {
++        return ColorType.getColorType(type);
++    }
++
++    /**
++     * Gets the {@link ColorType} of this {@link ItemStack}s-{@link Material}
++     * @return the color of this {@link ItemStack}s-Material, or <i>other</i> if it has no supported color
++     */
++    @NotNull
++    public ColorType getColorTypeOrOther(@NotNull ColorType other) {
++        ColorType output = getColorType();
++        return output != null ? output : other;
++    }
++
++    /**
++     * Gets the {@link ColorType} of this {@link ItemStack}s-{@link Material}
++     * @return the color of this {@link ItemStack}s-{@link Material}, or {@link ColorType#WHITE} if it has no supported color
++     */
++    @NotNull
++    public ColorType getColorTypeOrWhite() {
++        return getColorTypeOrOther(ColorType.WHITE);
++    }
++
++    /**
++     * Changes the Material-{@link ColorType} of this {@link ItemStack} using the {@link Material#getDyeableMaterial()}
++     * @param newColor the new color, which the {@link Material} will become
++     * @throws IllegalArgumentException if the {@link ItemStack} {@link Material} has no {@link DyeableMaterial}-group
++     */
++    public void changeColor(@NotNull ColorType newColor) throws IllegalArgumentException {
```

Comment:
Unsure about this; Material provides the method and this doesn't explicitly say it uses `setType`